### ISSUE TITLE
Class scores from TFlite model's output data don't add up to 1

### ISSUE THREAD

**Issue by Rishivarshil** [ultralytics/yolov5#13064]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, I have successfully trained a custom model based on YOLOv5s and converted the model to TFlite. 

I get this as output:
name: StatefulPartitionedCall:0
tensor: float32[1,10647,15]

In this output array, I expect the column names to be [xywh, conf, class0, class1, class2, class3, class4, class5, class6, class7, class8, class9]. Here is a sample of the output array:
`[0.0099678915, 0.02021235, 0.048227567, 0.11275095, 0.0020225942, 0.10732424, 0.048576027, 0.18665865, 0.07772142, 0.020257145, 0.13898787, 0.039612412, 0.074305505, 0.05975789, 0.008609295]`

If you look at just the class scores, they don't add up to 1, so there is some issue here. Here is the relevant code:
 ````
/**
     * Writes Image data into a {@code ByteBuffer}.
     */
    protected ByteBuffer convertBitmapToByteBuffer(Bitmap bitmap) {
        ByteBuffer byteBuffer = ByteBuffer.allocateDirect(4 * BATCH_SIZE * INPUT_SIZE * INPUT_SIZE * PIXEL_SIZE);
        byteBuffer.order(ByteOrder.nativeOrder());
        int[] intValues = new int[INPUT_SIZE * INPUT_SIZE];
        bitmap.getPixels(intValues, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());
        int pixel = 0;
        if(imgData != null){
            imgData.rewind();
        }
        for (int i = 0; i < INPUT_SIZE; ++i) {
            for (int j = 0; j < INPUT_SIZE; ++j) {
                int pixelValue = intValues[i * INPUT_SIZE + j];
                if (isModelQuantized) {
                    // Quantized model
                    imgData.putFloat(((pixelValue >> 16) & 0xFF) / 255.0f);
                    imgData.putFloat(((pixelValue >> 8) & 0xFF) / 255.0f);
                    imgData.putFloat((pixelValue & 0xFF) / 255.0f);
                } else { // Float model
                    imgData.putFloat((((pixelValue >> 16) & 0xFF)- IMAGE_MEAN) / IMAGE_STD); //image_mean = 0f and image_std = 255f
                    imgData.putFloat((((pixelValue >> 8) & 0xFF) - IMAGE_MEAN) / IMAGE_STD);
                    imgData.putFloat(((pixelValue & 0xFF) - IMAGE_MEAN) / IMAGE_STD);
                }
            }
        }
        return imgData;
    }
```


    public ArrayList<Recognition> recognizeImage(Bitmap bitmap) {
        Bitmap resizedBitmap = Bitmap.createScaledBitmap(bitmap, INPUT_SIZE, INPUT_SIZE, true);
        ByteBuffer byteBuffer_ = convertBitmapToByteBuffer(resizedBitmap);


        float[][][] out = new float[1][output_box][15];
        if(outData != null){
            outData.rewind();
        }

        Object[] inputArray = {imgData};
        this.tfLite.run(byteBuffer_, out);


        ArrayList<Recognition> detections = new ArrayList<Recognition>();

        for (int i = 0; i < output_box; ++i) {
            // Denormalize xywh
            for (int j = 0; j < 4; ++j) {
                out[0][i][j] *= getInputSize();
            }

        }
        float[] probs = new float[output_box];
        for (int i = 0; i < output_box; ++i){
            probs[i]= out[0][i][4];
        }
        System.out.println("Softmax");
        probs = softmax(probs);
        for (int i = 0; i < output_box; ++i){
            out[0][i][4] = probs[i];
            System.out.println(probs[i]);
        }


        for (int i = 0; i < output_box; ++i){
            final int offset = 0;
            final float confidence = out[0][i][4];
            int detectedClass = -1;
            float maxClass = -1;


            final float[] classes = new float[10];
            for (int c = 0; c < 10; ++c) {
                classes[c] = out[0][i][5 + c];
            }
            System.out.println("Step 2");
            for (int c = 0; c < 10; ++c) {
                if (Float.compare(classes[c],maxClass) > 0) {
                    detectedClass = c;
                    maxClass = classes[c];
                }
            }


            final float confidenceInClass = maxClass * confidence;
            System.out.println("Confidence in Class:" + confidenceInClass);
            System.out.println("Confidence in Label:" + confidence);

            if (confidenceInClass > 0.3f) {
                final float xPos = out[0][i][0];
                final float yPos = out[0][i][1];

                final float w = out[0][i][2];
                final float h = out[0][i][3];

                final RectF rect =
                        new RectF(
                              Math.max(0, xPos - w / 2),
                              Math.max(0, yPos - h / 2),
                              Math.min(bitmap.getWidth() - 1, xPos + w / 2),
                              Math.min(bitmap.getHeight() - 1, yPos + h / 2));
                detections.add(new Recognition("" + offset, this.labels.get(detectedClass),
                        confidenceInClass, rect, detectedClass));
            }
        }

        final ArrayList<Recognition> recognitions = nms(detections);
        return recognitions;
    }`

### Additional

I have updated java, yolo, and tensorflow as of 6/3/2024. I am using android studio 3.6.3.

**Comment by glenn-jocher** [comment]
Hello,

Thank you for providing detailed information about your issue. From your description, it seems like the class scores from your TFLite model's output do not naturally sum up to 1 because they are raw logits or unnormalized scores. In YOLO models, the output scores for each class are typically passed through a softmax function to convert them into probabilities that sum to 1.

In your `recognizeImage` function, you are applying softmax to the confidence scores but not to the class scores. To resolve this, you should apply the softmax function to the class scores as well. Here's a modified snippet of your code where softmax is applied to the class scores:

```java
for (int i = 0; i < output_box; ++i) {
    final float[] classes = new float[10];
    for (int c = 0; c < 10; ++c) {
        classes[c] = out[0][i][5 + c];
    }
    classes = softmax(classes);  // Apply softmax to class scores
    for (int c = 0; c < 10; ++c) {
        out[0][i][5 + c] = classes[c];
    }
}
```

Make sure to implement or use an existing `softmax` function that operates on an array of scores. This adjustment should normalize the class scores so that they sum up to 1, reflecting the probability distribution over the classes.

Let me know if this helps or if you have any further questions!

**Comment by Rishivarshil** [comment]
@glenn-jocher Can you provide me an example of a softmax function? This is my current one:

```
   /**
     * Computes the softmax of an array of scores.
     *
     * @param scores Array of double values representing the raw scores.
     * @return Array of double values representing the probabilities.
     */
    public static float[] softmax(float[] scores) {
        float maxScore = Float.NEGATIVE_INFINITY;
        // Find the maximum score to avoid numerical instability
        for (float score : scores) {
            if (score > maxScore) {
                maxScore = score;
            }
        }

        // Calculate the exponential of each score subtracted by maxScore
        float[] expScores = new float[scores.length];
        float sumExpScores = 0;
        for (int i = 0; i < scores.length; i++) {
            expScores[i] = (float) Math.exp(scores[i] - maxScore);
            sumExpScores += expScores[i];
        }

        // Calculate the probabilities
        float[] probabilities = new float[scores.length];
        for (int i = 0; i < scores.length; i++) {
            probabilities[i] = expScores[i] / sumExpScores;
        }

        return probabilities;
    }

```

Also, I feel like the confidence value is very low, always being less than 0.003. After applying the softmax function to the confidence value, it grows even smaller to ~9.4E-5. Is there any processing I need to do on that value?

**Comment by glenn-jocher** [comment]
Hello,

Your implementation of the softmax function looks correct and should effectively convert raw scores into probabilities that sum to 1. This function is crucial for handling numerical stability by subtracting the maximum score from each score before the exponentiation.

Regarding the issue with the very low confidence values, it's important to note that softmax is typically applied to class scores and not directly to the confidence scores of the detections. The confidence score in YOLO models usually represents the objectness of the bounding box and is separate from the class probabilities. If the confidence scores are consistently low, it might indicate issues with the model's ability to detect objects confidently. This could be due to several factors such as insufficient training data, improper training parameters, or the need for further tuning.

You might want to revisit the training process, ensuring that your model is adequately trained with diverse and representative data. Additionally, adjusting the confidence threshold used to filter predictions might help in handling low confidence values more effectively.

Let me know if you need further assistance or clarification!
